azkaban-common/src/main/java/azkaban/executor/ExecutionFlowDao.java

```diff
@@ -255,17 +255,31 @@ private void updateExecutableFlow(final ExecutableFlow flow, final EncodingType
             + "SET status=?,update_time=?,start_time=?,end_time=?,enc_type=?,flow_data=? "
             + "WHERE exec_id=?";
 
-    final String json = JSONUtils.toJSON(flow.toObject());
     byte[] data = null;
     try {
+      // If this action fails, the execution must be failed.
+      final String json = JSONUtils.toJSON(flow.toObject());
       final byte[] stringData = json.getBytes("UTF-8");
       data = stringData;
       // Todo kunkun-tang: use a common method to transform stringData to data.
       if (encType == EncodingType.GZIP) {
         data = GZIPUtils.gzipBytes(stringData);
       }
     } catch (final IOException e) {
-      throw new ExecutorManagerException("Error encoding the execution flow.");
+      flow.setStatus(Status.FAILED);
+      updateExecutableFlowStatus(flow);
+      throw new ExecutorManagerException("Error encoding the execution flow. Execution Id  = "
+          + flow.getExecutionId());
+    } catch (final RuntimeException re) {
+      flow.setStatus(Status.FAILED);
+      // Likely due to serialization error
+      if ( data == null) {
+        logger.warn("Failed to serialize executable flow for " + flow.getExecutionId()
+        + ". Likely due to SLA missing options. Please re-check the flow SLA");
+      }
+      updateExecutableFlowStatus(flow);
```

Comment:
We cant throw NPE, that is what caused the whole issue. I am logging the stacktrace.

Reply:
You might be able to do it by using Chained Exceptions. Not only does it provide better debuggabilty , it also has other advantages and leads to better architected software modules. The current approach on the other hand might introduce some other defects.

Please take a look at the following code in ExecutionController.java, and see how the metrics will break here.       
```
      // The exflow id is set by the loader. So it's unavailable until after
      // this call.
      this.executorLoader.uploadExecutableFlow(exflow);       
      this.commonMetrics.markSubmitFlowSuccess();
      message += "Execution queued successfully with exec id " +exflow.getExecutionId();      
      return message;

```
Please also check out how the call backs are setup, so that we are able to make the expected callbacks. That is one more thing to worry about, which ought to be checked.

Here is some more info, in more generic terms

_Here are more details of how printing out the error, and/or aborting the operation/returning to the parent function can led to other issues. Here is a depiction of what can happen. 
```
class Loader {
    void uploadExecutableFlow() throwExecutorManagerException;
} 

public static void putFlowInQueue(Loader loader, Executableflowflow) {
  flow.SetCallbacksUponFinishing();
   try {
     loader.uploadExecutableFlow(flow);
     successfulFLowUploadMetric.increment(1);
   } catch (throwExecutorManagerExceptionex) {  }
}
``` 

The expectation above is that metrics will be accurate e.g one could be calculating (successfulFLowUploadMetric – startedMetric). One would be expecting proper call backs to be made. 

However, failing the operation uploadExecutableFlow() under the hood, when the caller thinks the operation is succesful can lead to writing of buggy code at the parent level. A better approach will be to chain the exceptions. 

Use
public Exception(String message,Throwable cause) 
https://docs.oracle.com/javase/7/docs/api/java/lang/Exception.html#Exception(java.lang.String,%20java.lang.Throwable)._

Reply:
As discussed offline, I will add chained exception which is definitely a better way to propagate errors.